### DISCUSSION TITLE
Assistance Required: Issues with Calling AuxVariable in GeneralVectorPostprocessor

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#27113] [Q&A General] [answered]
Dear moose experts,

I hope this message finds you well. Recently, I have been working on developing a custom VectorPostprocessor named FeatureMatePropVectorPostprocessor, which is derived through public inheritance from GeneralVectorPostprocessor. In this class, I need to access an AuxVariable, and to achieve this, I've attempted to initialize it in the constructor using the `coupledValue` function. However, during the compilation process, I encountered the following error:

```C++
In file included from /home/pw-moose/projects/yinglong/build/unity_src/vectorpostprocessors_Unity.C:3:
/home/pw-moose/projects/yinglong/src/vectorpostprocessors/FeatureMatePropVectorPostprocessor.C: In constructor 'FeatureMatePropVectorPostprocessor::FeatureMatePropVectorPostprocessor(const InputParameters&)':
/home/pw-moose/projects/yinglong/src/vectorpostprocessors/FeatureMatePropVectorPostprocessor.C:17:21: error: 'coupledValue' was not declared in this scope; did you mean 'coupledScalar'?
   17 |     _energy_elastic(coupledValue("energy_elastic")),
      |                     ^~~~~~~~~~~~
      |                     coupledScalar
make: *** [/home/pw-moose/projects/moose/framework/build.mk:151: /home/pw-moose/projects/yinglong/build/unity_src/vectorpostprocessors_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1 
```

In an attempt to resolve this issue, I included the `Coupleable.h` header file to make the `CoupledValue()` function available, yet the problem persisted. Furthermore, following the compiler's suggestion, I tried using the `coupledScalarValue()` function. Unfortunately, this approach did not suit my needs as it failed to correctly call the AuxVariable during the simulation phase.

Given these challenges, I am seeking guidance on the correct method to call an AuxVariable within a VectorPostprocessor Object. Could you please advise on how to address this issue? Any suggestions or insights you could provide would be greatly appreciated.

Thank you for your time and assistance. I look forward to your valuable input.

Best regards,

Wei

**Comment by GiudGiud** [comment, marked as answer]
Hello

This vector postprocessor must not inherit the Coupleable interface (a base class to MANY objects, so often that we forget it has to be inherited).
What are the base classes for this VPP? What does it inherit from?
If not from Coupleable, then add `public Coupleable` to the header, and add it to the constructor too

Guillaume

**Reply by PengWei97** [reply]
Hi Guillaume,

Thank you very much for your guidance; the previous issue has been resolved. However, I've encountered another problem. The simulation indicates that the value of the auxvariable type requiring coupling must be based on nodes rather than elements. Afterwards, I converted the auxvariable type variable into the MaterialProperty type and used `getMaterialProperty` to call it within the VectorValuedProblem (VVP). Surprisingly, it compiled successfully, and the simulation passed.

Therefore, I believe the issue I encountered revolves around understanding why converting the auxVariable type quantity into the MaterialProperty type quantity leads to normal simulation. This seems to involve grasping the concept of [_qp].

Best wishes,
Wei

**Reply by GiudGiud** [reply]
`[_qp]` is indexing the property (or the variable value) at the index of the quadrature point (Qp) on the element (or side for a sideVPP). 
Properties and variable values are computed for all the Qps at once when considering an element, for efficiency reasons. Then you can pick the value you need from that array of size `number of Qps in the quadrature for an element (/side)`

**Reply by PengWei97** [reply]
Hi Guillaume,

Thank you for your insightful guidance. Additionally, after consulting some resources, I've gained a clearer understanding of the distinction between nodes and quadrature integration points. Here's my interpretation: Nodes are points located at the physical positions of an element, instrumental in defining the element's geometric shape and interpolation functions. Consequently, field variables (such as displacement, temperature, etc.) are approximated at these nodes. On the other hand, Gauss quadrature integration points serve as internal points for numerical integration, strategically distributed within the element to facilitate the calculation of integral properties (like components of the stiffness matrix and mass matrix), thus accurately assessing the element's contribution to the overarching problem.

Moreover, my understanding extends to the definition of MaterialProperty<T> type parameters, which seems to be inherently linked to [_qp]. This can be observed when outputting _qp in MaterialObject, which always outputs values in the range 0 to 4, when the TRI3 grid type and first-order Lagrangian shape function was used. I'm curious to know if my interpretation aligns with reality.

Finally, I'd like to delve deeper into whether it's possible to output parameters of the MaterialProperty<T> type as nodal. Specifically, I'm referring to the possibility of outputting values for the VuxVariable elastic_energy in AuxKernel::ElasticEnergyAux using the node type, rather than the element type. To my knowledge, AuxKernel::ElasticEnergyAux primarily outputs values based on the element type. I wonder if there's a way to adjust this to node-based outputs. If it is possible, how should ElasticEnergyAux be modified? If not, what is the reason?

Best wishes,
Wei

**Reply by GiudGiud** [reply]
Hello

> Finally, I'd like to delve deeper into whether it's possible to output parameters of the MaterialProperty type as nodal.

there are currently no nodal material properties, due to the ambiguity of which material definition to use when the node is in between two blocks. 

> Specifically, I'm referring to the possibility of outputting values for the VuxVariable elastic_energy in AuxKernel::ElasticEnergyAux using the node type, rather than the element type

so with an auxiliary variable, you CAN use nodal variables. But dont rely on using material properties at nodes to that.
I think what you may want to try if you MUST have a nodal variable is:
- use an elemental variable (for example a MONOMIAL variable) to convert the material property to an aux-variable
- use a ProjectionAux to project this elemental variable onto a nodal variable

Guillaume

**Reply by PengWei97** [reply]
Hi Guillaume,

Using ProjectionAux did not work, I used the following input settings,
```bash 
[AuxVariables]
  [./f_elastic_elem]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./f_elastic_node]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[AuxKernels]
  [./f_elastic_elem]
    type = ElasticEnergyAux
    variable = f_elastic_elem
    execute_on = TIMESTEP_BEGIN
  [../]
  [./f_elastic_node]
    type = ProjectionAux
    v = f_elastic_elem # Variable to take the value of.
    variable = f_elastic_node
    execute_on = TIMESTEP_END
  [../]
[]
```

ProjectionAux did not successfully map f_elastic_elem to f_elastic_node.

Wei

**Reply by GiudGiud** [reply]
how different are they? what do they look like?

**Reply by PengWei97** [reply]
![image](https://github.com/idaholab/moose/assets/56183775/41495c32-ff3b-4fcf-ad59-509fa574962c)
f_elastic_elem is as expected, but f_elastic_node is always 0.

**Reply by GiudGiud** [reply]
Interesting. which timestep are we seeing in this picture?
I ll run a quick test locally
EDIT: elem to nodal should work.

Are you sure we are not watching the result at timestep_begin? What is the output execute_on flag?

**Reply by PengWei97** [reply]
The timestep is step 03, and the output module is executed at timestep_end.
```bash 
[Outputs]
  execute_on = 'timestep_end'
  [my_exodus]
    type = Nemesis
    file_base = ./${my_filename}/out_${my_filename}
  [../]
[]
```

By the way, even if two auxkernels are set to run at timestep begin or end, the results will be the same as the one just uploaded.

**Reply by GiudGiud** [reply]
ok weird.
This does not seem right. Is there maybe an `inactive` or `active` parameter set for auxkernels that would prevent the kernel from executing? Do you see it executing with Debug/show_execution_order=ALWAYS?

Is there a material property output or a transfer that could be overwriting the auxvariable value? or another auxkernel?

Guillaume

**Reply by PengWei97** [reply]
Hi, this problem has been solved. In ProjectionAux, we need to set `use_displaced_mesh = false`, although I don't know why.

Wei

**Reply by GiudGiud** [reply]
Interesting. That seems like a bug. i ll investigate

**Reply by GiudGiud** [reply]
It is a bug. Raised this issue.
https://github.com/idaholab/moose/issues/27134

thanks for investigating it

**Reply by GiudGiud** [reply]
The fix should be in this PR 
https://github.com/idaholab/moose/pull/27144
